File: clients/csharp/Model/CityType.cs
namespace Komus24.Model
{
    /// <summary>
    /// TODO - Document
    /// </summary>
    public abstract class CityType
    {
        /// <summary> Write CityType to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read CityType from reader </summary>
        public static CityType ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case Manhattan.TAG:
                    return Manhattan.ReadFrom(reader);
                case Inline.TAG:
                    return Inline.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// TODO - Document
        /// </summary>
        public class Manhattan : CityType
        {
            public const int TAG = 0;
        
            /// <summary>
            /// TODO - Document
            /// </summary>
            public Komus24.Model.Vec2Int Size { get; set; }
            /// <summary>
            /// TODO - Document
            /// </summary>
            public Komus24.Model.Vec2Int BlockSize { get; set; }
            /// <summary>
            /// TODO - Document
            /// </summary>
            public int Refills { get; set; }
        
            public Manhattan() { }
        
            public Manhattan(Komus24.Model.Vec2Int size, Komus24.Model.Vec2Int blockSize, int refills)
            {
                this.Size = size;
                this.BlockSize = blockSize;
                this.Refills = refills;
            }
        
            /// <summary> Read Manhattan from reader </summary>
            public static new Manhattan ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Manhattan();
                result.Size = Komus24.Model.Vec2Int.ReadFrom(reader);
                result.BlockSize = Komus24.Model.Vec2Int.ReadFrom(reader);
                result.Refills = reader.ReadInt32();
                return result;
            }
        
            /// <summary> Write Manhattan to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                Size.WriteTo(writer);
                BlockSize.WriteTo(writer);
                writer.Write(Refills);
            }
        
            /// <summary> Get string representation of Manhattan </summary>
            public override string ToString() {
                string stringResult = "Manhattan { ";
                stringResult += "Size: ";
                stringResult += Size.ToString();
                stringResult += ", ";
                stringResult += "BlockSize: ";
                stringResult += BlockSize.ToString();
                stringResult += ", ";
                stringResult += "Refills: ";
                stringResult += Refills.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// TODO - Document
        /// </summary>
        public class Inline : CityType
        {
            public const int TAG = 1;
        
            /// <summary>
            /// TODO - Document
            /// </summary>
            public string[] Cells { get; set; }
        
            public Inline() { }
        
            public Inline(string[] cells)
            {
                this.Cells = cells;
            }
        
            /// <summary> Read Inline from reader </summary>
            public static new Inline ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Inline();
                result.Cells = new string[reader.ReadInt32()];
                for (int cellsIndex = 0; cellsIndex < result.Cells.Length; cellsIndex++)
                {
                    result.Cells[cellsIndex] = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(reader.ReadInt32()));
                }
                return result;
            }
        
            /// <summary> Write Inline to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Cells.Length);
                foreach (var cellsElement in Cells)
                {
                    var cellsElementData = System.Text.Encoding.UTF8.GetBytes(cellsElement);
                    writer.Write(cellsElementData.Length);
                    writer.Write(cellsElementData);
                }
            }
        
            /// <summary> Get string representation of Inline </summary>
            public override string ToString() {
                string stringResult = "Inline { ";
                stringResult += "Cells: ";
                stringResult += "[ ";
                int cellsIndex = 0;
                foreach (var cellsElement in Cells)
                {
                    if (cellsIndex != 0) {
                        stringResult += ", ";
                    }
                    stringResult += "\"" + cellsElement + "\"";
                    cellsIndex++;
                }
                stringResult += " ]";
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: clients/go/model/city_type.go
package model

import "fmt"
import "io"
import . "komus24/stream"

// TODO - Document
type CityType interface {
    // Write CityType to writer
    Write(writer io.Writer)

    // Get string representation of CityType
    String() string
}

// Read CityType from reader
func ReadCityType(reader io.Reader) CityType {
    switch ReadInt32(reader) {
    case 0:
        return ReadCityTypeManhattan(reader)
    case 1:
        return ReadCityTypeInline(reader)
    }
    panic("Unexpected tag value")
}

// TODO - Document
type CityTypeManhattan struct {
    // TODO - Document
    Size Vec2Int32
    // TODO - Document
    BlockSize Vec2Int32
    // TODO - Document
    Refills int32
}

func NewCityTypeManhattan(size Vec2Int32, blockSize Vec2Int32, refills int32) CityTypeManhattan {
    return CityTypeManhattan {
        Size: size,
        BlockSize: blockSize,
        Refills: refills,
    }
}

// Read Manhattan from reader
func ReadCityTypeManhattan(reader io.Reader) CityTypeManhattan {
    var size Vec2Int32
    size = ReadVec2Int32(reader)
    var blockSize Vec2Int32
    blockSize = ReadVec2Int32(reader)
    var refills int32
    refills = ReadInt32(reader)
    return CityTypeManhattan {
        Size: size,
        BlockSize: blockSize,
        Refills: refills,
    }
}

// Write Manhattan to writer
func (cityTypeManhattan CityTypeManhattan) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    size := cityTypeManhattan.Size
    size.Write(writer)
    blockSize := cityTypeManhattan.BlockSize
    blockSize.Write(writer)
    refills := cityTypeManhattan.Refills
    WriteInt32(writer, refills)
}

// Get string representation of Manhattan
func (cityTypeManhattan CityTypeManhattan) String() string {
    stringResult := "{ "
    stringResult += "Size: "
    size := cityTypeManhattan.Size
    stringResult += size.String()
    stringResult += ", "
    stringResult += "BlockSize: "
    blockSize := cityTypeManhattan.BlockSize
    stringResult += blockSize.String()
    stringResult += ", "
    stringResult += "Refills: "
    refills := cityTypeManhattan.Refills
    stringResult += fmt.Sprint(refills)
    stringResult += " }"
    return stringResult
}

// TODO - Document
type CityTypeInline struct {
    // TODO - Document
    Cells []string
}

func NewCityTypeInline(cells []string) CityTypeInline {
    return CityTypeInline {
        Cells: cells,
    }
}

// Read Inline from reader
func ReadCityTypeInline(reader io.Reader) CityTypeInline {
    var cells []string
    cells = make([]string, ReadInt32(reader))
    for cellsIndex := range cells {
        var cellsElement string
        cellsElement = ReadString(reader)
        cells[cellsIndex] = cellsElement
    }
    return CityTypeInline {
        Cells: cells,
    }
}

// Write Inline to writer
func (cityTypeInline CityTypeInline) Write(writer io.Writer) {
    WriteInt32(writer, 1)
    cells := cityTypeInline.Cells
    WriteInt32(writer, int32(len(cells)))
    for _, cellsElement := range cells {
        WriteString(writer, cellsElement)
    }
}

// Get string representation of Inline
func (cityTypeInline CityTypeInline) String() string {
    stringResult := "{ "
    stringResult += "Cells: "
    cells := cityTypeInline.Cells
    stringResult += "[ "
    for cellsIndex, cellsElement := range cells {
        if cellsIndex != 0 {
            stringResult += ", "
        }
        stringResult += "\"" + cellsElement + "\""
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}